pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -110,27 +121,124 @@ type manager struct {
+	if err != nil {
+		klog.Errorf("[memorymanager] could not initialize checkpoint manager: %v, please drain node and remove policy state file", err)
+		return err
+	}
+	m.state = stateImpl
+
+	err = m.policy.Start(m.state)
+	if err != nil {
+		klog.Errorf("[memorymanager] policy start error: %v", err)
+		return err
+	}
+
+	if m.policy.Name() == string(mmpolicy.TypeNone) {
+		return nil
+	}
+
+	// TODO: we should to decide if we really need the reconcile loop under the memory management
+	// go wait.Until(func() { m.reconcileState() }, m.reconcilePeriod, wait.NeverStop)
 	return nil
 }
 
-// AddContainer saves the value of requested memory for the guranteed pod under the state and set memory affinity according to the topolgy manager
-func (m *manager) AddContainer(p *v1.Pod, c *v1.Container, containerID string) error {
+// AddContainer saves the value of requested memory for the guaranteed pod under the state and set memory affinity according to the topolgy manager
+func (m *manager) AddContainer(pod *v1.Pod, container *v1.Container, containerID string) error {
+	m.Lock()
+	// Get memory blocks assigned to the container during Allocate()
+	blocks := m.state.GetMemoryBlocks(string(pod.UID), container.Name)
+	m.Unlock()
+
+	if len(blocks) > 0 {
+		// it does not possible that memory blocks under the same container will have different NUMA affinity,
+		// so we get NUMA node from the first memory block
+		numaNode := strconv.Itoa(blocks[0].NUMAAffinity)
+		err := m.containerRuntime.UpdateContainerResources(containerID, &runtimeapi.LinuxContainerResources{CpusetMems: numaNode})
+		if err != nil {
+			klog.Errorf("[memorymanager] AddContainer error: error updating cpuset.mems for container (pod: %s, container: %s, container id: %s, err: %v)", pod.Name, container.Name, containerID, err)
+
+			m.Lock()
+			err = m.policyRemoveContainerByRef(string(pod.UID), container.Name)
+			if err != nil {
+				klog.Errorf("[memorymanager] AddContainer rollback state error: %v", err)
+			}
+			m.Unlock()
+		}
+		return err
+	}
+
+	klog.V(5).Infof("[memorymanager] update container resources is skipped due to memory blocks are empty")
 	return nil
 }
 
 // Allocate is called to pre-allocate memory resources during Pod admission.
 func (m *manager) Allocate(pod *v1.Pod, container *v1.Container) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// Call down into the policy to assign this container CPUs if required.
+	err := m.policy.Allocate(m.state, pod, container)
+	if err != nil {
+		klog.Errorf("[memorymanager] Allocate error: %v", err)
+		return err
+	}
 	return nil
 }
 
 // RemoveContainer removes the container from the state
 func (m *manager) RemoveContainer(containerID string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// if error appears it means container entry already does not exist under the container map
+	podUID, containerName, err := m.containerMap.GetContainerRef(containerID)
+	if err != nil {
+		return nil
+	}
```

Comment:
I think the reason why cpu manager return nil here is they call RemoveContainer twice in internal container life cycle.
but we call Remove Contaienr only once, so I think we should return err here.
But I'm not sure what will happen if RemoveContainer returns err during container deletion.